### PR TITLE
DEX-1215 Add Prometheus Monitoring

### PR DESCRIPTION
**Pull request by bluemellophone**
This PR adds Prometheus monitoring at `/metrics` for the Houston app.

---

This PR does several things to allow for convenient Prometheus monitoring.  The most notable is adding a new blueprint to serve the `prometheus_client` information at `/metrics`.  

# Celery Background Compute

To aid in updating the metrics in the background and recording the update in the foreground web process, there is a new `[POST] /api/v1/prometheus/` endpoint that allows a background worker to post to the running web server as the default Oauth2 internal user.  The JSON data in the request is then used to update the in-memory Prometheus `Counter` and `Gauge` objects in the foreground process. 

# Model Registering

To add a model to the Prometheus monitoring (e.g., provide automatic counting stats from the database), you simply need to use `app.extensions.prometheus.register_prometheus_model(cls)`.  This design is similar to the Elasticsearch registering process.

# Metric Definitions

```python
info = Info(
    'info',
    'Description of the Houston environment',
)

models = Gauge(
    'models',
    'Number of total models by type',
    ['cls'],
)

taxonomies = Gauge(
    'taxonomies',
    'Number of taxonomies by metric',
    ['taxonomy', 'metric'],
)

logins = Gauge(
    'logins',
    'Number of unique users who have logged in over the past specified days',
    ['days'],
)

requests = Counter(
    'requests',
    'Number of total requests by endpoint since start',
    ['method', 'endpoint'],
)

responses = Counter(
    'responses',
    'Number of total responses by endpoint since start',
    ['method', 'endpoint', 'code'],
)

tasks_ = Gauge(
    'tasks',
    'Number of Celery tasks by function',
    ['function'],
)
```

# Example Metric Scrape Output

``` text 
# HELP python_gc_objects_collected_total Objects collected during gc
# TYPE python_gc_objects_collected_total counter
python_gc_objects_collected_total{generation="0"} 23379.0
python_gc_objects_collected_total{generation="1"} 7370.0
python_gc_objects_collected_total{generation="2"} 521.0
# HELP python_gc_objects_uncollectable_total Uncollectable object found during GC
# TYPE python_gc_objects_uncollectable_total counter
python_gc_objects_uncollectable_total{generation="0"} 0.0
python_gc_objects_uncollectable_total{generation="1"} 0.0
python_gc_objects_uncollectable_total{generation="2"} 0.0
# HELP python_gc_collections_total Number of times this generation was collected
# TYPE python_gc_collections_total counter
python_gc_collections_total{generation="0"} 711.0
python_gc_collections_total{generation="1"} 64.0
python_gc_collections_total{generation="2"} 5.0
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="3",minor="9",patchlevel="2",version="3.9.2"} 1.0
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 6.47139328e+08
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 1.84590336e+08
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.65843923266e+09
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 2.7199999999999998
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 17.0
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 1.048576e+06
# HELP info_info Description of the Houston environment
# TYPE info_info gauge
info_info{full_version="0.1.0.6ab6c56a.6ab6c56a5706de0c977f216ffec15325a7cbd9fc",git_revision="6ab6c56a5706de0c977f216ffec15325a7cbd9fc",version="0.1.0.6ab6c56a"} 1.0
# HELP models Number of total models by type
# TYPE models gauge
models{cls="app.modules.annotations.models.Annotation"} 0.0
models{cls="app.modules.asset_groups.models.AssetGroupSighting"} 0.0
models{cls="app.modules.asset_groups.models.AssetGroup"} 0.0
models{cls="app.modules.assets.models.Asset"} 0.0
models{cls="app.modules.collaborations.models.Collaboration"} 0.0
models{cls="app.modules.encounters.models.Encounter"} 0.0
models{cls="app.modules.individuals.models.Individual"} 0.0
models{cls="app.modules.names.models.Name"} 0.0
models{cls="app.modules.organizations.models.Organization"} 0.0
models{cls="app.modules.relationships.models.Relationship"} 0.0
models{cls="app.modules.sightings.models.Sighting"} 0.0
models{cls="app.modules.social_groups.models.SocialGroup"} 0.0
models{cls="app.modules.users.models.User"} 2.0
# HELP taxonomies Number of taxonomies by metric
# TYPE taxonomies gauge
taxonomies{metric="total",taxonomy="None"} 0.0
# HELP logins Number of unique users who have logged in over the past specified days
# TYPE logins gauge
logins{days="1"} 2.0
logins{days="7"} 2.0
logins{days="30"} 2.0
logins{days="90"} 2.0
logins{days="180"} 2.0
logins{days="365"} 2.0
logins{days="None"} 2.0
# HELP requests_total Number of total requests by endpoint since start
# TYPE requests_total counter
requests_total{endpoint="/api/v1/site-settings/heartbeat",method="GET"} 12.0
requests_total{endpoint="/api/v1/prometheus/",method="POST"} 3.0
requests_total{endpoint="/metrics",method="GET"} 3.0
requests_total{endpoint="/api/v1/auth/tokens",method="POST"} 1.0
# HELP requests_created Number of total requests by endpoint since start
# TYPE requests_created gauge
requests_created{endpoint="/api/v1/site-settings/heartbeat",method="GET"} 1.6584392371832402e+09
requests_created{endpoint="/api/v1/prometheus/",method="POST"} 1.6584392481135283e+09
requests_created{endpoint="/metrics",method="GET"} 1.658439330083453e+09
requests_created{endpoint="/api/v1/auth/tokens",method="POST"} 1.6584393513652084e+09
# HELP responses_total Number of total responses by endpoint since start
# TYPE responses_total counter
responses_total{code="200",endpoint="/api/v1/site-settings/heartbeat",method="GET"} 12.0
responses_total{code="200",endpoint="/api/v1/prometheus/",method="POST"} 3.0
responses_total{code="200",endpoint="/metrics",method="GET"} 2.0
responses_total{code="200",endpoint="/api/v1/auth/tokens",method="POST"} 1.0
# HELP responses_created Number of total responses by endpoint since start
# TYPE responses_created gauge
responses_created{code="200",endpoint="/api/v1/site-settings/heartbeat",method="GET"} 1.6584392371834145e+09
responses_created{code="200",endpoint="/api/v1/prometheus/",method="POST"} 1.6584392481201918e+09
responses_created{code="200",endpoint="/metrics",method="GET"} 1.6584393300878098e+09
responses_created{code="200",endpoint="/api/v1/auth/tokens",method="POST"} 1.6584393513962204e+09
# HELP tasks Number of Celery tasks by function
# TYPE tasks gauge
tasks{function="None"} 58.0
tasks{function="app.extensions.elasticsearch.tasks.es_task_index_bulk"} 11.0
tasks{function="app.extensions.intelligent_agent.tasks.twitterbot_collect"} 40.0
tasks{function="app.extensions.sage.tasks.sage_task_jobs_sync"} 4.0
tasks{function="app.extensions.prometheus.tasks.prometheus_update"} 2.0
tasks{function="app.extensions.tus.tasks.tus_task_cleanup"} 1.0
```

### CODE REVIEW

File: app/modules/encounters/models.py [hunk 1]
```diff
@@ -165,6 +165,23 @@ def get_taxonomy_guid(self, sighting_fallback=True):
             taxonomy_guid = self.sighting.get_taxonomy_guid()
         return taxonomy_guid
 
+    def get_taxonomy_names(self, **kwargs):
```
Comment (hwindsor) [inline app/modules/encounters/models.py:168]: Have a look in the Taxonomy class, doesn't get_all_names give you this?

Comment (karenc) [reply app/modules/encounters/models.py:185]: Thanks!  I'm going to look into it

Comment (karenc) [reply app/modules/encounters/models.py:185]: Does it look right?

```python
return Taxonomy(taxonomy_guid).get_all_names()
```

**Comment by karenc** [comment]
I added tests and tested with the zebra db:

```
models{cls="app.modules.annotations.models.Annotation"} 77631.0
models{cls="app.modules.asset_groups.models.AssetGroupSighting"} 0.0
models{cls="app.modules.asset_groups.models.AssetGroup"} 26725.0
models{cls="app.modules.assets.models.Asset"} 54054.0
models{cls="app.modules.collaborations.models.Collaboration"} 0.0
models{cls="app.modules.encounters.models.Encounter"} 66714.0
models{cls="app.modules.individuals.models.Individual"} 14090.0
models{cls="app.modules.names.models.Name"} 14517.0
models{cls="app.modules.organizations.models.Organization"} 0.0
models{cls="app.modules.relationships.models.Relationship"} 760.0
models{cls="app.modules.sightings.models.Sighting"} 0.0
models{cls="app.modules.social_groups.models.SocialGroup"} 0.0
models{cls="app.modules.users.models.User"} 53.0
```

File: app/modules/encounters/models.py (hunk 1 above)
Comment (hwindsor) [reply app/modules/encounters/models.py:185]: Looks a lot simpler, as long as it does the same thing (which is what I was less certain about) then that's great

Comment (karenc) [reply app/modules/encounters/models.py:185]: Yeah, I wasn't sure either.  That's why I added a test to check the output is still the same

**Review by bluemellophone — COMMENTED** [review]
LGTM, self-review

**Pull request merged**